file_writer.go

```diff
@@ -176,6 +201,28 @@ func (f *FileWriter) SetDeadline(t time.Time) error {
 // of this, it is important that Close is called after all data has been
 // written.
 func (f *FileWriter) Write(b []byte) (int, error) {
+	// Encrypt data chunk if file in HDFS encrypted zone.
+	if f.enc != nil && len(b) > 0 {
+		var offset int
+		for offset < len(b) {
+			size := min(len(b)-offset, aesChunkSize)
+			ciphertext, err := aesCtrStep(f.offset, f.enc, b[offset:offset+size])
+			if err != nil {
+				return offset, err
+			}
+			writtenSize, err := f.writeImpl(ciphertext)
+			offset += writtenSize
```

Comment:
I think we're violating the contract of `io.Writer` by returning the output size, not the amount of bites of the input successfully written. I could be misunderstanding, though.

Reply:
with encryption, we return `offset`, and `offset` updated only by `f.writeImpl` output.

Reply:
Exactly. io.Writer says:

> Write writes len(p) bytes from p to the underlying data stream. It returns the number of bytes written from p (0 <= n <= len(p)) and any error encountered that caused the write to stop early.

If the length of the written ciphertext is longer than the plaintext, then we return n > len(p). If it's less, then we return n < len(p) without an error.

My rudimentary understanding of block ciphers (confirmed by [this SO answer](https://stackoverflow.com/a/3717552) is that the ciphertext is usually, but not always, the same length as the plaintext. My guess is that the tests are passing because of an alignment between our block size and the input, but it should be possible to construct a pathological case that breaks that assumption.

Reply:
Btw, it's not just the interface contract that could break here: if `writtenSize` is larger than `size`, we will skip over `writtenSize - size` bytes in `b`.